### ISSUE TITLE
PDB Reading adds gaps between a number break in structure, even with guess_atoms

### ISSUE THREAD

**Issue by jadolfbr** [MDAnalysis/mdanalysis#4164]
## Expected behavior ##

Load PDB, everything is connected unless TER statements are there. Or a different way to infer connectivity such as bond lengths or angles.  

## Actual behavior ##

Load PDB, if the numbering goes from 6-8, such is the case when modeling homologous proteins like antibodies, they are not connected as they should be.  PHI and PSI dihedral angles are blank as the API thinks that there is a literal structural gap between them.  

## Code to reproduce the behavior ##

``` python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD,  GRO, PDB, TPR, XTC, TRR,  PRMncdf, NCDF

u = mda.Universe("renumbered.pdb")
for i, res in enumerate(u.residues):
    print(i, res, res.phi_selection())

....

```

## Current version of MDAnalysis ##

- Which version are you using? 2.5.0
- Which version of Python? 3.9.16
- Which operating system? Mac

## PDB Snippet of a few residues that show example:
```
ATOM    426  N   PHE A  60       5.126  70.037  32.435  1.00 41.47
ATOM    427  CA  PHE A  60       6.181  69.239  33.083  1.00 41.38
ATOM    428  C   PHE A  60       6.710  68.231  32.061  1.00 41.82
ATOM    429  O   PHE A  60       5.921  67.644  31.322  1.00 41.62
ATOM    430  CB  PHE A  60       5.614  68.441  34.269  1.00 43.26
ATOM    431  CG  PHE A  60       5.833  69.068  35.622  1.00 42.97
ATOM    432  CD1 PHE A  60       4.796  69.742  36.271  1.00 47.37
ATOM    433  CD2 PHE A  60       7.052  68.936  36.273  1.00 43.77
ATOM    434  CE1 PHE A  60       4.983  70.302  37.547  1.00 46.35
ATOM    435  CE2 PHE A  60       7.254  69.494  37.533  1.00 45.53
ATOM    436  CZ  PHE A  60       6.213  70.178  38.176  1.00 45.87
ATOM    437  N   PRO A  61       8.036  68.023  31.989  1.00 41.48
ATOM    438  CA  PRO A  61       8.490  66.958  31.081  1.00 41.09
ATOM    439  C   PRO A  61       7.944  65.604  31.504  1.00 41.06
ATOM    440  O   PRO A  61       7.972  65.291  32.680  1.00 42.08
ATOM    441  CB  PRO A  61      10.013  67.011  31.222  1.00 40.90
ATOM    442  CG  PRO A  61      10.287  68.431  31.623  1.00 40.52
ATOM    443  CD  PRO A  61       9.174  68.717  32.607  1.00 41.10
ATOM    444  N   ALA A  62       7.410  64.822  30.569  1.00 41.67
ATOM    445  CA  ALA A  62       6.819  63.524  30.921  1.00 40.99
ATOM    446  C   ALA A  62       6.708  62.559  29.737  1.00 40.96
ATOM    447  O   ALA A  62       6.970  62.923  28.587  1.00 39.89
ATOM    448  CB  ALA A  62       5.458  63.729  31.557  1.00 40.19
ATOM    449  N   LEU A  63       6.328  61.322  30.049  1.00 41.61
ATOM    450  CA  LEU A  63       6.127  60.281  29.055  1.00 43.01
ATOM    451  C   LEU A  63       4.686  59.762  29.122  1.00 43.91
ATOM    452  O   LEU A  63       4.378  58.858  29.897  1.00 44.37
ATOM    453  CB  LEU A  63       7.121  59.139  29.282  1.00 42.92
ATOM    454  CG  LEU A  63       8.598  59.451  29.009  1.00 42.69
ATOM    455  CD1 LEU A  63       9.495  58.385  29.635  1.00 43.02
ATOM    456  CD2 LEU A  63       8.877  59.562  27.523  1.00 41.72
ATOM    457  N   LEU A  64       3.818  60.333  28.287  1.00 44.93
ATOM    458  CA  LEU A  64       2.376  60.042  28.312  1.00 45.31
ATOM    459  C   LEU A  64       2.052  58.828  27.443  1.00 45.15
ATOM    460  O   LEU A  64       2.472  58.757  26.289  1.00 45.26
ATOM    461  CB  LEU A  64       1.585  61.271  27.837  1.00 45.50
ATOM    462  CG  LEU A  64       1.927  62.576  28.577  1.00 45.74
ATOM    463  CD1 LEU A  64       1.286  63.788  27.920  1.00 44.62
ATOM    464  CD2 LEU A  64       1.523  62.481  30.043  1.00 45.72
ATOM    465  N   LYS A  65       1.302  57.884  28.007  1.00 45.98
ATOM    466  CA  LYS A  65       1.111  56.557  27.403  1.00 46.89
ATOM    467  C   LYS A  65       0.210  56.575  26.173  1.00 47.73
ATOM    468  O   LYS A  65      -0.651  57.445  26.046  1.00 49.23
ATOM    469  N   VAL A  71       2.378  56.273  23.781  1.00 51.28
ATOM    470  CA  VAL A  71       3.632  56.626  24.449  1.00 51.12
ATOM    471  C   VAL A  71       4.337  57.771  23.713  1.00 50.92
ATOM    472  O   VAL A  71       4.823  57.580  22.596  1.00 52.01
ATOM    473  CB  VAL A  71       4.587  55.423  24.536  1.00 50.84
ATOM    474  CG1 VAL A  71       5.672  55.692  25.578  1.00 50.57
ATOM    475  CG2 VAL A  71       3.814  54.156  24.881  1.00 51.62
ATOM    476  N   VAL A  72       4.380  58.952  24.331  1.00 49.18
ATOM    477  CA  VAL A  72       4.954  60.145  23.689  1.00 47.94
ATOM    478  C   VAL A  72       5.714  60.992  24.707  1.00 46.41
ATOM    479  O   VAL A  72       5.374  60.987  25.891  1.00 46.61
ATOM    480  CB  VAL A  72       3.856  61.017  23.029  1.00 48.51
ATOM    481  CG1 VAL A  72       4.475  62.121  22.163  1.00 49.46
ATOM    482  CG2 VAL A  72       2.914  60.160  22.189  1.00 49.02
ATOM    483  N   ALA A  73       6.737  61.710  24.239  1.00 44.16
ATOM    484  CA  ALA A  73       7.578  62.547  25.103  1.00 42.81
ATOM    485  C   ALA A  73       7.122  64.007  25.084  1.00 41.46
ATOM    486  O   ALA A  73       7.363  64.730  24.114  1.00 41.10
ATOM    487  CB  ALA A  73       9.037  62.455  24.675  1.00 41.81
ATOM    488  N   ARG A  74       6.470  64.437  26.158  1.00 39.41
ATOM    489  CA  ARG A  74       6.103  65.837  26.307  1.00 38.86
ATOM    490  C   ARG A  74       7.291  66.589  26.892  1.00 38.25
ATOM    491  O   ARG A  74       7.901  66.133  27.860  1.00 38.77
ATOM    492  CB  ARG A  74       4.871  65.994  27.204  1.00 38.25
ATOM    493  CG  ARG A  74       4.407  67.437  27.378  1.00 37.34
ATOM    494  CD  ARG A  74       3.231  67.530  28.337  1.00 36.92
ATOM    495  NE  ARG A  74       3.631  67.322  29.724  1.00 37.09
ATOM    496  CZ  ARG A  74       2.793  67.079  30.730  1.00 35.64
ATOM    497  NH1 ARG A  74       1.480  67.010  30.533  1.00 36.06
ATOM    498  NH2 ARG A  74       3.275  66.892  31.946  1.00 32.62
ATOM    499  N   ASN A  75       7.622  67.720  26.270  1.00 37.51
ATOM    500  CA  ASN A  75       8.638  68.656  26.747  1.00 37.20
ATOM    501  C   ASN A  75       9.896  67.943  27.215  1.00 37.09
ATOM    502  O   ASN A  75      10.348  68.120  28.339  1.00 37.87
ATOM    503  CB  ASN A  75       8.044  69.558  27.844  1.00 37.15
ATOM    504  CG  ASN A  75       8.909  70.780  28.158  1.00 37.78
ATOM    505  OD1 ASN A  75       8.715  71.441  29.186  1.00 41.57
ATOM    506  ND2 ASN A  75       9.858  71.085  27.286  1.00 33.73
ATOM    507  N   GLY A  76      10.440  67.118  26.329  1.00 37.56
ATOM    508  CA  GLY A  76      11.702  66.422  26.567  1.00 37.17
ATOM    509  C   GLY A  76      11.576  64.990  27.051  1.00 36.85
ATOM    510  O   GLY A  76      12.552  64.248  27.035  1.00 37.59
ATOM    511  N   GLY A  77      10.395  64.590  27.504  1.00 36.50
ATOM    512  CA  GLY A  77      10.235  63.268  28.110  1.00 36.44
ATOM    513  C   GLY A  77      10.762  63.212  29.538  1.00 35.67
ATOM    514  O   GLY A  77      10.008  62.942  30.462  1.00 35.95
ATOM    515  N   VAL A  78      12.061  63.451  29.709  1.00 34.91
ATOM    516  CA  VAL A  78      12.690  63.501  31.034  1.00 33.85
ATOM    517  C   VAL A  78      13.152  64.931  31.328  1.00 32.41
ATOM    518  O   VAL A  78      13.505  65.667  30.415  1.00 30.66
ATOM    519  CB  VAL A  78      13.891  62.515  31.160  1.00 34.00
ATOM    520  CG1 VAL A  78      13.399  61.106  31.307  1.00 36.58
ATOM    521  CG2 VAL A  78      14.842  62.608  29.967  1.00 34.30
```

**Comment by hmacdope** [comment]
@jadolfbr sorry for the delay.

@MDAnalysis/coredevs what do we think here, could we add perhaps a `non-sequential=...` kwarg that allows the selection to search not for the next resiude by number as is done [here](https://github.com/MDAnalysis/mdanalysis/blob/a14e98df322c675e9ed7555c11547d55797552e3/package/MDAnalysis/core/topologyattrs.py#L822C10-L822C10)

**Comment by udaysahu09** [comment]
Fix Issue #4164: Support non-sequential residue numbering in dihedral selection methods

When loading PDB files with non-sequential residue numbering (e.g., residues
60-65 followed by 71-78), dihedral selection methods failed to find adjacent
residues, returning None instead of valid AtomGroup selections.

The root cause was using residue.resid instead of residue.ix to determine
adjacent residues in the chain. When PDB numbering is non-sequential, 
residue.resid ± 1 may not exist in the structure.

Solution: Use chain position (residue.ix ± 1) to find neighbors, then 
validate with resid checks. Fall back to explicit resid search if needed.
This works for both sequential and non-sequential numbering.

Modified 8 methods in topologyattrs.py:
- phi_selection() / phi_selections()
- psi_selection() / psi_selections()  
- omega_selection() / omega_selections()
- _get_next_residues_by_resid() / _get_prev_residues_by_resid()

Fixes #4164
- Maintains backward compatibility
- Supports non-sequential PDB numbering (antibody CDRs, etc.)
- All existing tests pass"

**Comment by jadolfbr** [comment]
Why is this spam when the issue hasn’t been looked into for 2-3 years? Yes
Claude code and codex will help fix issues.  Please look over this fix,
review, and merge.  This is a simple bug.

**Comment by BradyAJohnston** [comment]
@jadolfbr I marked this as spam because it is the description for a PR, but commented on the issue instead of creating the PR itself. This happened on another issue as well from the same user, and there was no sign of any actual PR. It seemed like it was likely low-quality LLM slop that was being copied and pasted, incorrectly, into the issue instead of creating a PR and adding the description there so I figured it was not worth trying to engage with.

**Comment by jadolfbr** [comment]
Ah ok, wow.  Thanks for the info and checking into these throughly. I
didn’t realize things like that were happening.

**Comment by IAlibay** [comment]
I just wanted to chime in here in support of @BradyAJohnston 's decision. The MDAnalysis developers, and frankly the entire open source ecosystem, is being enormously stretched by the current surge in LLM usage. Unfortunately this leads to situations where we are having to make decisions such as these - we don't enjoy them, but we can't also afford to try to engage with everything.

I would encourage everyone to read our AI policy. It explains where we currently stand on the issue and serves as a good starting point for further discussions.

**Comment by BradyAJohnston** [comment]
To comment on the actual issue / feature request: I think having a kwarg to allow searching for "non-sequential" resids would be a good solution. Should be simple enough to implement if we like the idea of it. Would be potential for issues if a structure had larger missing regions that are often common in un-modelled regions of structures. Might want to have a check for distance if two residues aren't sequential?

**Comment by IAlibay** [comment]
I'm wondering if this is something that would be dealt with once that gemmi parser is complete?

**Comment by BradyAJohnston** [comment]
I don't think the parser will make any difference, because we'll still be reading the disjointed residue numbers. We wouldn't want to make any assumptions during parsing, so just a potential analysis kwarg that instead allows just using the _next available_ residue rather than the numerically next would be the approach?
